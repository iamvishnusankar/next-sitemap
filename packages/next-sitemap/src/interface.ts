type MaybeUndefined<T> = T | undefined
type MaybePromise<T> = T | Promise<T>

export interface IRobotPolicy {
  userAgent: string
  disallow?: string | string[]
  allow?: string | string[]
}

export interface IRobotsTxt {
  policies?: IRobotPolicy[]
  additionalSitemaps?: string[]
}

export interface IConfig {
  siteUrl: string
  changefreq: string
  priority: any
  sitemapBaseFileName?: string
  sourceDir?: string
  outDir?: string
  sitemapSize?: number
  generateRobotsTxt: boolean
  robotsTxtOptions?: IRobotsTxt
  autoLastmod?: boolean
  exclude?: string[]
  alternateRefs?: Array<AlternateRef>
  transform?: (
    config: IConfig,
    url: string
  ) => MaybePromise<MaybeUndefined<ISitemapField>>
  additionalPaths?: (
    config: AdditionalPathsConfig
  ) => MaybePromise<MaybeUndefined<ISitemapField>[]>
  trailingSlash?: boolean
}

export type AdditionalPathsConfig = Readonly<
  IConfig & {
    transform: NonNullable<IConfig['transform']>
  }
>

export interface IBuildManifest {
  pages: {
    [key: string]: string[]
  }
}

export interface IPreRenderManifest {
  routes: {
    [key: string]: any
  }
}

export interface IExportMarker {
  exportTrailingSlash: boolean
}

export interface INextManifest {
  build: IBuildManifest
  preRender?: IPreRenderManifest
}

export interface ISitemapChunk {
  path: string
  fields: ISitemapField[]
  filename: string
}

export interface IRuntimePaths {
  BUILD_MANIFEST: string
  PRERENDER_MANIFEST: string
  SITEMAP_FILE: string
  ROBOTS_TXT_FILE: string
  EXPORT_MARKER: string
}

export type AlternateRef = {
  href: string
  hreflang: string
}

export type ISitemapField = {
  loc: string
  lastmod?: string
  changefreq?: string
  priority?: number
  alternateRefs?: Array<AlternateRef>
  news?: News
}

export interface News {
  publication: NewsPublication
  publication_date: string
  title: string
}

export interface NewsPublication {
  name: string
  language: string
}
