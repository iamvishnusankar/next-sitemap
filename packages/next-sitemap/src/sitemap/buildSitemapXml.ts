import { AlternateRef, ISitemapField, News } from '../interface'
import { withXMLTemplate } from './withXMLTemplate'

export const buildSitemapXml = (fields: ISitemapField[]): string => {
  const content = fields
    .map((fieldData) => {
      const field: Array<string> = []

      // Iterate all object keys and key value pair to field-set
      for (const key of Object.keys(fieldData)) {
        if (!fieldData[key]) {
          continue
        }

        if (key === 'alternateRefs') {
          field.push(buildAlternateRefsXml(fieldData.alternateRefs))
          continue
        }

        if (key === 'news') {
          field.push(buildNewsXml(fieldData.news))
          continue
        }

        field.push(`<${key}>${fieldData[key]}</${key}>`)
      }

      // Append previous value and return
      return `<url>${field.join('')}</url>\n`
    })
    .join('')

  return withXMLTemplate(content)
}

export const buildAlternateRefsXml = (
  alternateRefs: Array<AlternateRef> = []
): string => {
  return alternateRefs
    .map((alternateRef) => {
      return `<xhtml:link rel="alternate" hreflang="${alternateRef.hreflang}" href="${alternateRef.href}"/>`
    })
    .join('')
}

export const buildNewsXml = (news?: News): string => {
  if (!news) {
    return ''
  }
  return `
     <news:news>
        <news:publication>
          <news:name>${news.publication.name}</news:name>
          <news:language>${news.publication.language}</news:language>
        </news:publication>
        <news:publication_date>${news.publication_date}</news:publication_date>
        <news:title>${news.title}</news:title>
      </news:news>
  `
}
