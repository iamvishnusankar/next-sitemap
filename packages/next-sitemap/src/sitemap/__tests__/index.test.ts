import { ISitemapField } from '../../interface'
import { buildSitemapXml } from '../buildSitemapXml'

describe('buildSitemapXml', () => {
  test('snapshot test to exclude undefined values from final sitemap', () => {
    // Sample fields
    const fields: ISitemapField[] = [
      {
        loc: 'https://example.com',
        lastmod: undefined,
      },
      {
        loc: 'https://example-news.com',
        news: undefined,
      },
      {
        loc: 'https://example.com',
        lastmod: 'some-value',
        alternateRefs: [
          {
            href: 'https://example.com/en',
            hreflang: 'en',
          },
          {
            href: 'https://example.com/fr',
            hreflang: 'fr',
          },
        ],
      },
    ]

    // Generate sitemap
    const sitemap = buildSitemapXml(fields)

    // Expect the generated sitemap to match snapshot.
    expect(sitemap).toMatchSnapshot()
  })
  test('snapshot test for handling the google news from final sitemap', () => {
    // Sample fields
    const fields: ISitemapField[] = [
      {
        loc: 'https://example-ny-times.com',
        lastmod: '2021-03-05T20:38:12Z',
        news: {
          publication: {
            name: 'The New York Times',
            language: 'en-US',
          },
          publication_date: '2019-09-30T08:00:03Z',
          title:
            "Lesson of the Day: 'Kehinde Wiley’s Times Square Monument: That’s No Robert E. Lee'",
        },
      },
    ]

    // Generate sitemap
    const sitemap = buildSitemapXml(fields)

    // Expect the generated sitemap to match snapshot.
    expect(sitemap).toMatchSnapshot()
  })
})
