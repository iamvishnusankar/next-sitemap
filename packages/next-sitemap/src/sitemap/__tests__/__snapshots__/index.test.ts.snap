// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildSitemapXml snapshot test for handling the google news from final sitemap 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:mobile=\\"http://www.google.com/schemas/sitemap-mobile/1.0\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\">
<url><loc>https://example-ny-times.com</loc><lastmod>2021-03-05T20:38:12Z</lastmod>
     <news:news>
        <news:publication>
          <news:name>The New York Times</news:name>
          <news:language>en-US</news:language>
        </news:publication>
        <news:publication_date>2019-09-30T08:00:03Z</news:publication_date>
        <news:title>Lesson of the Day: 'Kehinde Wiley’s Times Square Monument: That’s No Robert E. Lee'</news:title>
      </news:news>
  </url>
</urlset>"
`;

exports[`buildSitemapXml snapshot test to exclude undefined values from final sitemap 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:mobile=\\"http://www.google.com/schemas/sitemap-mobile/1.0\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\">
<url><loc>https://example.com</loc></url>
<url><loc>https://example-news.com</loc></url>
<url><loc>https://example.com</loc><lastmod>some-value</lastmod><xhtml:link rel=\\"alternate\\" hreflang=\\"en\\" href=\\"https://example.com/en\\"/><xhtml:link rel=\\"alternate\\" hreflang=\\"fr\\" href=\\"https://example.com/fr\\"/></url>
</urlset>"
`;
